sw/device/tests/dif/dif_keymgr_unittest.cc

```diff
@@ -0,0 +1,1114 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_keymgr.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "keymgr_regs.h"  // Generated
+
+namespace dif_keymgr_unittest {
+namespace {
+
+/**
+ * Returns a `uint32_t` with a single zero bit.
+ */
+uint32_t AllOnesExcept(uint32_t index) { return ~(1u << index); }
+
+/**
+ * Returns a vector of values for a given enum type.
+ *
+ * Assumes that the values are sequential and the first value is 0. `last` must
+ * be the last valid value for the given enum and is included in the returned
+ * vector.
+ */
+template <typename T>
+std::vector<T> CreateEnumVector(T last) {
+  using TT = typename std::underlying_type<T>::type;
+  std::vector<T> res;
+  for (TT i = 0; i <= static_cast<TT>(last); ++i) {
+    res.push_back(static_cast<T>(i));
+  }
+  return res;
+}
+
+/**
+ * Returns a seemingly valid, i.e. nonzero, pointer for the given type or a
+ * `nullptr`.
+ */
+template <typename T>
+T *GetGoodBadPtrArg(bool is_good) {
+  if (is_good) {
+    return reinterpret_cast<T *>(0x1);
```

Comment:
Nit: This should probably be `alignof(T)`. It's UB to create misaligned pointers in C++ (which can be confusing because Rust has no such rule...)

Reply:
Thanks for catching this!